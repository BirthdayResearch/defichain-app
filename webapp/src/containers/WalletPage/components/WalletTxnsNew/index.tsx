import React, { useState, useEffect } from 'react';
import {
  Card,
  Table,
  CardBody,
  FormGroup,
  CustomInput,
  Row,
  Col,
} from 'reactstrap';
import { connect } from 'react-redux';
import {
  MdArrowUpward,
  MdArrowDownward,
  MdCompareArrows,
} from 'react-icons/md';
import styles from './WalletTxns.module.scss';
import { I18n } from 'react-redux-i18n';
import {
  fetchWalletTxnsRequest,
  fetchWalletTokenTransactionsListRequestLoading,
  fetchWalletTokenTransactionsListRequestPaginationLoading,
} from '../../reducer';
import {
  RECIEVE_CATEGORY_LABEL,
  SENT_CATEGORY_LABEL,
  WALLET_TXN_PAGE_SIZE,
} from '../../../../constants';
import Pagination from '../../../../components/Pagination';
import { numberWithCommas } from '../../../../utils/utility';
import cloneDeep from 'lodash/cloneDeep';
import EllipsisText from 'react-ellipsis-text';

interface WalletTxnsProps {
  unit: string;
  fetchWalletTxns: (
    currentPage: number,
    pageSize: number,
    intialLoad?: boolean
  ) => void;
  tokenSymbol?: string;
  tokenAddress?: string;
  fetchWalletTokenTransactionsListRequestLoading: (
    currentPage: number,
    symbol?: string,
    owner?: string
  ) => void;
  fetchWalletTokenTransactionsListRequestPaginationLoading: (
    currentPage: number,
    symbol?: string,
    owner?: string
  ) => void;
  data: any[];
  isLoading: boolean;
  isError: string;
  stop: boolean;
}

const WalletTxns: React.FunctionComponent<WalletTxnsProps> = (
  props: WalletTxnsProps
) => {
  const {
    fetchWalletTokenTransactionsListRequestLoading,
    fetchWalletTokenTransactionsListRequestPaginationLoading,
    tokenAddress,
    tokenSymbol,
    data,
    isLoading,
    isError,
    stop,
  } = props;
  const [currentPage, setCurrentPage] = useState(1);
  const [tableRows, setTableRows] = useState<any[]>([]);
  const [tableData, setTableData] = useState<any>([]);
  const [includeRewards, setIncludeRewards] = useState(false);
  const pageSize = WALLET_TXN_PAGE_SIZE;
  const total = tableData.length;
  const pagesCount = Math.ceil(total / pageSize);
  const to = (currentPage - 1) * pageSize + 1;
  const from = Math.min(total, currentPage * pageSize);

  useEffect(() => {
    fetchWalletTokenTransactionsListRequestLoading(
      currentPage,
      tokenSymbol,
      tokenAddress
    );
  }, []);

  const fetchData = (pageNum, disableReq = false) => {
    if (!disableReq) {
      if (pageNum > 1 && pageNum * pageSize > tableData.length) {
        return fetchWalletTokenTransactionsListRequestPaginationLoading(
          currentPage,
          tokenSymbol,
          tokenAddress
        );
      }
    }
    const newCloneTableData = cloneDeep(tableData);
    const rows = newCloneTableData.slice(
      (pageNum - 1) * pageSize,
      pageNum * pageSize
    );
    setTableRows(rows);
    return setCurrentPage(pageNum);
  };

  useEffect(() => {
    setTableData(
      data.filter((item) => {
        if (!includeRewards) return item.category !== 'Rewards';
        return !!item.category;
      })
    );
  }, [data, includeRewards]);

  useEffect(() => {
    fetchData(currentPage, true);
  }, [tableData]);

  const getTxnsTypeIcon = (type: string) => {
    if (type === SENT_CATEGORY_LABEL) {
      return <MdArrowUpward className={styles.typeIcon} />;
    }
    if (type === RECIEVE_CATEGORY_LABEL || type === 'Rewards') {
      return <MdArrowDownward className={styles.typeIconDownward} />;
    }
    if (type === 'PoolSwap')
      return <MdCompareArrows className={styles.typeIcon} />;
    return '';
  };

  const getAmountClass = (type: string) => {
    if (
      type === RECIEVE_CATEGORY_LABEL ||
      type === 'Rewards' ||
      type === 'PoolSwap'
    ) {
      return styles.colorGreen;
    }
    return '';
  };

  const walletTxnList = () => {
    if (isLoading)
      return <div>{I18n.t('containers.wallet.walletPage.loading')}</div>;
    if (isError) return <div>{isError}</div>;
    if (!tableData.length)
      return (
        <Card className='table-responsive-md'>
          <CardBody>
            {I18n.t('containers.wallet.walletTxns.noTransactions')}
          </CardBody>
        </Card>
      );
    return (
      <>
        <Card className={`${styles.card} table-responsive-md`}>
          <Table className={styles.table}>
            <tbody>
              {tableRows.map((item, id) => (
                <tr key={`${currentPage}-${id}`}>
                  <td>{getTxnsTypeIcon(item.category)}</td>
                  <td>
                    <div>{item.category}</div>
                    <div className={styles.unit}>{item.blockData.time}</div>
                  </td>
                  <td>
                    <div className={styles.txidvalue}>
                      <EllipsisText text={item.txid} length={60} />
                    </div>
                  </td>
                  <td
                    className={`text-right ${getAmountClass(item.category)}`}
                  >{`${numberWithCommas(item.amount)} ${item.symbolKey}`}</td>
                </tr>
              ))}
            </tbody>
          </Table>
        </Card>
        <Pagination
          label={I18n.t('containers.wallet.walletPage.paginationRange', {
            to,
            total,
            from,
          })}
          currentPage={currentPage}
          pagesCount={pagesCount}
          handlePageClick={fetchData}
          showNextOnly
          disableNext={stop}
        />
      </>
    );
  };

  return (
    <section className='mb-5'>
      <div className={styles.container}>
        <h2>{I18n.t('containers.wallet.walletPage.transactions')}</h2>
        <FormGroup>
          <CustomInput
            type='checkbox'
            id='includeRewards'
            label={I18n.t('containers.wallet.walletPage.includeRewards')}
            checked={includeRewards}
            onChange={() => {
              setIncludeRewards(!includeRewards);
            }}
          />
        </FormGroup>
      </div>
      <Row>
        <Col xs='12'>{walletTxnList()}</Col>
      </Row>
    </section>
  );
};

const mapStateToProps = (state) => {
  const {
    settings,
    wallet: {
      listAccountHistoryData: { isLoading, data, isError, stop },
    },
  } = state;
  return {
    unit: settings.appConfig.unit,
    isLoading,
    data,
    isError,
    stop,
  };
};

const mapDispatchToProps = {
  fetchWalletTxns: (currentPage, pageSize, intialLoad) =>
    fetchWalletTxnsRequest({ currentPage, pageSize, intialLoad }),
  fetchWalletTokenTransactionsListRequestLoading: (
    currentPage: number,
    symbol?: string,
    owner?: string
  ) =>
    fetchWalletTokenTransactionsListRequestLoading({
      symbol,
      owner,
      currentPage,
    }),
  fetchWalletTokenTransactionsListRequestPaginationLoading: (
    currentPage: number,
    symbol?: string,
    owner?: string
  ) =>
    fetchWalletTokenTransactionsListRequestPaginationLoading({
      symbol,
      owner,
      currentPage,
    }),
};

export default connect(mapStateToProps, mapDispatchToProps)(WalletTxns);
