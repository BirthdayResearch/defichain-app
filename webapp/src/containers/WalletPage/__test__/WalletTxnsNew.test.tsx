import React from 'react';
import WalletTxnsNew from '../components/WalletTxnsNew';
import { Provider } from 'react-redux';
import store from '../../../app/rootStore';
import Enzyme from '../../../utils/testUtils/enzyme';

describe('WalletTxnsNew component', () => {
  it('should check for snapshot', () => {
    const tokenSymbol = 'DFI';
    const wrapper = Enzyme.mount(
      <Provider store={store}>
        <WalletTxnsNew tokenSymbol={tokenSymbol} />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
