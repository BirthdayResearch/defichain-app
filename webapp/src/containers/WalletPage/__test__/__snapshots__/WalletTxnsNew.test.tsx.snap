// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletTxnsNew component should check for snapshot 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(WalletTxns)
    tokenSymbol="DFI"
  >
    <WalletTxns
      accountHistoryCount=""
      accountHistoryCountRequest={[Function]}
      blockCount={0}
      combineAccountHistoryData={
        Object {
          "data": Array [],
          "isError": "",
          "isLoading": false,
        }
      }
      data={Array []}
      fetchBlockCount={[Function]}
      fetchBlockDataForTrxRequestLoading={[Function]}
      fetchWalletTokenTransactionsListRequestLoading={[Function]}
      fetchWalletTokenTransactionsListResetRequest={[Function]}
      fetchWalletTxns={[Function]}
      isError="Invalid configuration"
      isLoading={false}
      minBlockHeight={0}
      tokenSymbol="DFI"
      unit="DFI"
      walletTxnCount={0}
      walletTxns={Array []}
    >
      <section
        className="mb-5"
      >
        <div>
          <h2>
            Transactions
          </h2>
          <div
            className="btn-group"
          >
            <FormGroup
              tag="div"
            >
              <div
                className="form-group"
              >
                <CustomInput
                  checked={false}
                  id="includeRewards"
                  label="Include reward transactions"
                  onChange={[Function]}
                  type="checkbox"
                >
                  <div
                    className="custom-checkbox custom-control"
                    hidden={false}
                  >
                    <input
                      checked={false}
                      className="custom-control-input"
                      id="includeRewards"
                      onChange={[Function]}
                      type="checkbox"
                    />
                    <label
                      className="custom-control-label"
                      htmlFor="includeRewards"
                    >
                      Include reward transactions
                    </label>
                  </div>
                </CustomInput>
              </div>
            </FormGroup>
            <Button
              color="link"
              onClick={[Function]}
              size="sm"
              tag="button"
            >
              <button
                aria-label={null}
                className="btn btn-link btn-sm"
                onClick={[Function]}
                type="button"
              >
                <MdFileDownload>
                  <IconBase
                    attr={
                      Object {
                        "viewBox": "0 0 24 24",
                      }
                    }
                  >
                    <svg
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      strokeWidth="0"
                      style={
                        Object {
                          "color": undefined,
                        }
                      }
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M19 9h-4V3H9v6H5l7 7 7-7zM5 18v2h14v-2H5z"
                        key="0"
                      />
                    </svg>
                  </IconBase>
                </MdFileDownload>
                <span
                  className="d-lg-inline"
                >
                  EXPORT DATA
                </span>
              </button>
            </Button>
            <DownloadCsvModal
              CsvModalOpen={false}
              downloadDisable={true}
              error=""
              filename="Transactions_2021-3-17_11-21-6.csv"
              handleCheckBox={[Function]}
              handleCsvButtonClick={[Function]}
              handleDownloadWindow={[Function]}
              handleRegularNumInputs={[Function]}
              maxBlock={[Function]}
              reqData={
                Object {
                  "blockHeight": 0,
                  "limit": 100,
                  "no_rewards": true,
                  "token": "DFI",
                }
              }
              toggle={[Function]}
              tokenSymbol="DFI"
              transactionData={Array []}
            >
              <Modal
                autoFocus={true}
                backdrop={true}
                backdropTransition={
                  Object {
                    "mountOnEnter": true,
                    "timeout": 150,
                  }
                }
                centered={true}
                container="body"
                fade={false}
                isOpen={false}
                keyboard={true}
                modalTransition={
                  Object {
                    "timeout": 300,
                  }
                }
                onClosed={[Function]}
                onOpened={[Function]}
                returnFocusAfterClose={true}
                role="dialog"
                scrollable={false}
                toggle={[Function]}
                trapFocus={false}
                unmountOnClose={true}
                zIndex={1050}
              />
            </DownloadCsvModal>
          </div>
        </div>
        <Row
          tag="div"
          widths={
            Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ]
          }
        >
          <div
            className="row"
          >
            <Col
              tag="div"
              widths={
                Array [
                  "xs",
                  "sm",
                  "md",
                  "lg",
                  "xl",
                ]
              }
              xs="12"
            >
              <div
                className="col-12"
              >
                <div>
                  Invalid configuration
                </div>
              </div>
            </Col>
          </div>
        </Row>
      </section>
    </WalletTxns>
  </Connect(WalletTxns)>
</Provider>
`;
